### ISSUE TITLE
Multiple TestRun implementations cause test report overwrites

### ISSUE THREAD

**Issue by aSemy** [kotest/kotest#4297]
Continued from https://github.com/kotest/kotest/pull/4291#issuecomment-2286872567

[Slack discussion](https://kotlinlang.slack.com/archives/CSYLDDZUY/p1724580427631109)

---

`:kotest-tests:kotest-tests-junitxml:check` is failing on CI.

I'm looking at fixing the failing `:kotest-tests:kotest-tests-junitxml` tests

https://github.com/kotest/kotest/pull/4291#issuecomment-2287047133

I think the problem is that Gradle permits creating multiple test runs. In KGP you do so like this:

```kotlin
kotlin {
   jvm {
      testRuns.named("main") {
         // the regular `./gradlew :jvmTest` task`
      }
      testRuns.create("jvmCustomTest") {
         // a custom `./gradlew :jvmCustomTest` task`
      }
   }
}
```

However, Kotest doesn't pick up on this, and so it outputs the reports for both runs to the same directory.

https://github.com/kotest/kotest/blob/408d5f0611a397df533119f23be9539721a915ca/kotest-extensions/kotest-extensions-junitxml/src/jvmMain/kotlin/io/kotest/extensions/junitxml/JunitXmlReporter.kt#L46

I'd like to be able to fix this automatically, so that the reports naturally get put into independent directories, but I'm not sure how.

Gradle launches Kotest using [KotestJunitPlatformTestEngine](https://github.com/kotest/kotest/blob/47f35f93c2fecda800e25795ad4adbbdc9e86da8/kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/KotestJunitPlatformTestEngine.kt#L27). Does Gradle report the task-specific report directory to KotestJunitPlatformTestEngine?

Alternatively, could Kotest pick up the name of the executed task in another way?

**Comment by sksamuel** [comment]
> Gradle launches Kotest using [KotestJunitPlatformTestEngine](https://github.com/kotest/kotest/blob/47f35f93c2fecda800e25795ad4adbbdc9e86da8/kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/KotestJunitPlatformTestEngine.kt#L27). Does Gradle report the task-specific report directory to KotestJunitPlatformTestEngine?

No, because gradle just listens to the test engine output and generates its own format.
The JunixXML output extension isn't really needed if we're executing via gradle, so I propose just deprecating it as a gradle feature (we can use it for people  who don't use gradle).
What do you think @aSemy

**Comment by sksamuel** [comment]
I think we can close this after marking JunitXmlReporter as deprecated ?

We don't need it anymore, given that the XML reports generated by Gradle are fully supported (and have been for a while, I know this ticket is old and we didn't in the past). And Gradle's reports are fully aware of multiple tasks (and specifically the new JvmTestSuite functionality).

I guess our implementation has slightly better nested test support, but I don't think that alone justifies the custom implementation, plus as @aSemy points out, it doesn't work with multiple runs (and will require hacks to do so - because in the the JUnit5 way, it's Gradle that is listening to events and writing them, it's not injecting the path location and letting JUnit do it).

Added tests that confirm that we support test suites.

Any objections?
